enhancements/two-nodes-openshift/2no.md

```diff
@@ -0,0 +1,510 @@
+* Minimize any differences to the primary OpenShift platforms
+* Avoid any decisions that would prevent future implementation and support for upgrade/downgrade paths between two-node and traditional architectures 
+* Provide an OpenShift cluster experience that is identical to that of a 3-node hyperconverged cluster, but with 2 nodes
+
+### Non-Goals
+
+* Workload resilience - see related [Pre-DRAFT enhancement](https://docs.google.com/document/d/1TDU_4I4LP6Z9_HugeC-kaQ297YvqVJQhBs06lRIC9m8/edit)
+* Resilient storage - see future enhancement
+* Support for platforms other than bare metal including automated ci testing
+* Support for other topologies (eg. hypershift)
+* Adding worker nodes
+* Creation RHEL-HA events and metrics for consumption by the OpenShift monitoring stack (Deferred to post-MVP)
+
+## Proposal
+
+Use the RHEL-HA stack (Corosync, and Pacemaker), which has been used to delivered supported 2-node cluster experiences for multiple decades, to manage cri-o, kubelet, and the etcd daemon.
+etcd will run as as a voting member on both nodes.
+We will take advantage of RHEL-HA's native support for systemd and re-use the standard cri-o and kublet units, as well as create a new Open Cluster Framework (OCF) script for etcd.
+The existing startup order of cri-o, then kubelet, then etcd will be preserved.
+The `etcdctl`, `etcd-metrics`, and `etcd-readyz` containers will remain part of the static pod.
+
+Use RedFish compatible Baseboard Management Controllers (BMCs) as our primary mechanism to power off (fence) an unreachable peer and ensure that it can do no harm while the remaining node continues.
+
+Upon a peer failure, the RHEL-HA components on the surivor will fence the peer and use the OCF script to restart etcd as a new cluster-of-one.
+
+Upon a network failure, the RHEL-HA components ensure that exactly one node will survive, fence it's peer, and use the OCF script to restart etcd as a new cluster-of-one.
+
+In both cases, the control-plane's dependance on etcd will cause it to respond with errors until etcd has been restarted.
+
+Upon rebooting, the RHEL-HA components ensure that a node remains inert (not running cri-o, kubelet, or etcd) until it sees it's peer.
+If the peer is likely to remain offline for an extended period of time, admin confirmation is required to allow the node to start OpenShift.
+
+When starting etcd, the OCF script will use etcd's cluster ID and version counter to determine whether the existing data directory can be reused, or must be erased before joining an active peer.
+
+### Workflow Description
+
+#### Cluster Creation
+
+Creation of a two node control-plane will be possible via the core installer (with an additional bootstrap node), and via the Assisted Installer (without an additional bootstrap node).
+
+In the case of the core OpenShift installer, the user-facing proceedure is unchanged from a standard "IPI" installation, other than the configuration of 2 nodes instead of 3.
```

Comment:
Lots of conversations on this topic... harking back to the point about not building something new, the way to minimise architectural differences on the installer front is to get the necessary bits for a "bootstrap + 2 node" path into the core installer, use the assisted installer to avoid the dedicated bootstrap node, and from there get it into the agent based installer.

Reply:
No requirement for this to be supported in the core installer, but AFAIU it needs to be implemented there (and is alleged to work already) in order to not invent something new in assisted.

Reply:
We can (and I think should) do validations at `create cluster` time to prevent anyone using it in the IPI flow.
Agree that we want a lot of the implementation to live inside the installer, just as most of how assisted does a regular installation is delegated to the installer.

Reply:
@mshitrit I think this thread can also be marked as resolved.